task_manager/task_manager.hh

```diff
@@ -0,0 +1,323 @@
+                return std::move(f).finally([module, id] {
+                    return sleep_abortable(module->get_task_ttl(), module->abort_source());
+                }).then_wrapped([module, id] (auto f) {
+                    f.ignore_ready_future();
+                    module->unregister_task(id);
+                });
+            });
+        }
+
+        static future<std::vector<task_id>> run_on_shards(task_ptr task, module_ptr module) {
+            return module->_tm.container().map([parent = seastar::make_foreign(task), module] (task_manager&) mutable -> foreign_task_ptr {
+                task_ptr child = parent->create_child(task_id::create_random_id(), std::move(parent)); // TODO Deexie the rest of fields
+                module->register_task(child);
+                child->start();
+                return seastar::make_foreign<task_ptr>(child);
+            }).then([task] (foreign_task_vector tasks) {
+                task->_children = std::move(tasks);
+                return make_ready_future<std::vector<task_id>>(boost::copy_range<std::vector<task_id>>(task->_children
+                    | boost::adaptors::transformed([] (foreign_task_ptr& child) { return child->_status.id; })));
+            });
+        }
+
+        future<> done() const noexcept {
+            return _done.get_shared_future();
+        }
+
+        void unregister() {
+            _module->unregister_task(_status.id);
+        }
+
+        void finish() noexcept {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::done;
+            _done.set_value();
+        }
+
+        void finish_failed(std::exception_ptr ex, std::string error) noexcept {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::failed;
+            _status.error = std::move(error);
+            _done.set_exception(ex);
+        }
+
+        void finish_failed(std::exception_ptr ex) noexcept {
+            finish_failed(ex, fmt::format("{}", ex));
+        }
+
+        virtual void abort() = 0;
+    };
+    class module {
+    private:
+        task_manager& _tm;
+        std::string _name;
+        task_map _tasks;
+        seastar::gate _gate;
+    public:
+        module(task_manager& tm, std::string name) : _tm(tm), _name(std::move(name)) {} // TODO Deexie protect from global usages.
+
+        static module_ptr make_module(task_manager& tm, std::string name) {
+            auto m = seastar::make_shared<module>(tm, name);
+            tm.register_module(std::move(name), m);
+            return m;
+        }
+
+        seastar::gate& async_gate() noexcept {
+            return _gate;
+        }
+
+        abort_source& abort_source() noexcept {
+            return _tm.abort_source();
+        }
+
+        std::chrono::seconds get_task_ttl() const {
+            return _tm.get_task_ttl();
```

Comment:
It would probably be better to expose a `get_task_manager()` accessor and let the caller just use it to get to to the task_manager methods.

We don't want to replicate all the task_manager methods in this class.